tokio/src/sync/mpsc/chan.rs

```diff
@@ -293,6 +295,79 @@ impl<T, S: Semaphore> Rx<T, S> {
         })
     }
 
+    /// Receives values into `buffer` up to its capacity
+    ///
+    /// If the buffer initially has 0 capacity, reserves `super::BLOCK_CAP` elements
+    pub(crate) fn recv_many(&mut self, cx: &mut Context<'_>, buffer: &mut Vec<T>) -> Poll<usize> {
+        use super::block::Read;
+        if buffer.capacity() == 0 {
+            buffer.reserve(super::BLOCK_CAP);
+        }
+        buffer.clear();
```

Comment:
I don't think we should clear the buffer. That's up to the user. I think it's better to support appending to the vector if the user wants that.

Reply:
There is a clear performance case to be made; particularly if this avoids copying elements in the caller.

Can you suggest the behavior if the buffer has already hit capacity at the time of the `recv_many` call?  `assert!(buffer.capacity() > buffer.len());` -- or `return Pending;` -- if there is a message that cannot be retrieved?

Along these lines, it may also make sense to eliminate `buffer.reserve(super::BLOCK_CAP);` -- I had included that behavior to make `recv_many` unfailing, but with this change the method can fail when the buffer's _full_ (len = capacity).  A zero-length buffer is -- in this sense -- always _full_.

Reply:
There are two options that make sense to me:

 * Add a third argument that specifies the maximum number of elements to add to the vector. In this case, the capacity is irrelevant, and we will resize the buffer if necessary.
 * Or use the capacity like this:

```rs
if buffer.len() == buffer.capacity() {
    buffer.reserve(super::BLOCK_CAP);
}
let max_number_added = buffer.capacity() - buffer.len();
```

In either case, I prefer that we do not clear the buffer.

Reply:
I adopted the second approach in the latest push.